File: worker_py/llm/openai.py
"""OpenAI-compatible Chat Completions adapter. Port of ``llm/openai.ts``.

Works with OpenAI and any compatible gateway via ``OPENAI_BASE_URL``.
"""

import os
from typing import Optional

from ..http import request_with_retry
from .base import LLMProvider


class OpenAILLM(LLMProvider):
    name = "openai"

    def __init__(self, api_key: str, model: Optional[str] = None, base_url: Optional[str] = None) -> None:
        self.api_key = api_key
        self.model = model or os.environ.get("OPENAI_MODEL") or "gpt-4o-mini"
        self.base_url = (base_url or os.environ.get("OPENAI_BASE_URL") or "https://api.openai.com/v1").rstrip("/")

    async def complete(
        self,
        prompt,
        *,
        system=None,
        max_tokens=1024,
        temperature=0.3,
        json=False,
        model=None,
        timeout_ms=30000,
    ) -> str:
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        body = {
            "model": model or self.model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": messages,
        }
        if json:
            body["response_format"] = {"type": "json_object"}

        res = await request_with_retry(
            "POST",
            f"{self.base_url}/chat/completions",
            headers={"Authorization": f"Bearer {self.api_key}", "content-type": "application/json"},
            json=body,
            timeout_ms=timeout_ms,
        )
        data = res.json()
        try:
            return str(data["choices"][0]["message"]["content"] or "").strip()
        except (KeyError, IndexError, TypeError):
            return ""

File: src/lib/core/llm/openai.ts
import { fetchWithRetry } from "../http";
import type { LLMCompleteOptions, LLMProvider } from "./types";

/**
 * OpenAI-compatible Chat Completions adapter. Works with OpenAI and any
 * compatible gateway (set OPENAI_BASE_URL), e.g. OpenRouter, Together, vLLM.
 */
export class OpenAILLM implements LLMProvider {
  readonly name = "openai";
  readonly model: string;
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(apiKey: string, model?: string, baseUrl?: string) {
    this.apiKey = apiKey;
    this.model = model || process.env.OPENAI_MODEL || "gpt-4o-mini";
    this.baseUrl = (baseUrl || process.env.OPENAI_BASE_URL || "https://api.openai.com/v1").replace(/\/$/, "");
  }

  async complete(prompt: string, opts: LLMCompleteOptions = {}): Promise<string> {
    const messages = [
      ...(opts.system ? [{ role: "system" as const, content: opts.system }] : []),
      { role: "user" as const, content: prompt },
    ];

    const res = await fetchWithRetry(
      `${this.baseUrl}/chat/completions`,
      {
        method: "POST",
        headers: { Authorization: `Bearer ${this.apiKey}`, "content-type": "application/json" },
        body: JSON.stringify({
          model: opts.model || this.model,
          max_tokens: opts.maxTokens ?? 1024,
          temperature: opts.temperature ?? 0.3,
          messages,
          ...(opts.json ? { response_format: { type: "json_object" } } : {}),
        }),
      },
      { timeoutMs: opts.timeoutMs ?? 30000 }
    );

    const json = await res.json();
    return String(json?.choices?.[0]?.message?.content ?? "").trim();
  }
}
